### ISSUE TITLE
Feature: support telescope.nvim's lsp.code_actions for ESLint actions

### ISSUE THREAD

**Issue by jose-elias-alvarez** [jose-elias-alvarez/nvim-lsp-ts-utils#8]
Telescope's `lsp.code_actions` [fetches actions directly from the language server](https://github.com/nvim-telescope/telescope.nvim/blob/253d3aaa6b43eac6b11341b325e34d37dc459af3/lua/telescope/builtin/lsp.lua#L114), making it impossible to inject ESLint code actions by overriding `vim.lsp.handlers["textDocument/codeAction"]` (which makes sense, since the code action handler is concerned with handling code actions, not getting them). 

I can think of two reasonable solutions and one not-so-reasonable but potentially less brittle solution:

1. Put together a snippet for a custom Telescope function that goes through the custom code action handler first (and otherwise replicates the default behavior) and host it on this repo's wiki. The only less-than-ideal part is that I'd need to keep the snippet updated if the underlying Telescope function changes to maintain feature parity and compliance with the Telescope API (which is also why I might not want to include it directly into the plugin).

2. See if the Telescope maintainers are open to making `lsp.code_actions` accept `results_lsp` as an optional argument (and continue to fetch the results if not passed in, so there would be no change in behavior for non-users). This is up to them and what they want for the plugin, but on a technical level it's pretty simple. It would still require setting up a custom function, but it would be simple enough that it could be easily included into the plugin.

The issue with those solutions is that this same problem would come up with any plugin that defines a code action handler, such as [lspsaga](https://github.com/glepnir/lspsaga.nvim), which leads me to...

3. Override `vim.lsp.buf_request` and `vim.lsp.buf_request_sync` so that they inject ESLint code actions into `textDocument/codeAction` requests (and pass them on to the original methods unmodified otherwise). This *seems* pretty straightforward, since Neovim doesn't care if you override them in `on_attach`, but I'm not sure if this is a good idea, and it would make bugs worse, since if the custom function doesn't work, most (or all?) LSP requests would stop working. 

Anyways, I'll give this some thought, and I'm happy to hear input from anyone else who happens to see this. (I'll probably put together a Telescope snippet for my own use and share it on the wiki either way.)

**Comment by VVKot** [comment]
@jose-elias-alvarez IMO option 3 sounds the best, precisely for the reasons you mention (there is lspsaga and potentially unbounded number of other plugins that would need it). Re: breaking LPS request by this logic - I guess it is fairly straight-forward to diagnose, and tests should cover most egregious bugs anyway?

On the broader point, current direction does not sit very well with me - but I don't know a better alternative. Expand diagnostics-ls to have "plugin points" for code action extractors? Run linters as separate language servers processes? I guess vscode's and coc's eslint solution are pretty much the latter (just not conforming to the LSP spec), so it actually might be a way to go.

**Comment by jose-elias-alvarez** [comment]
I share your unease and ultimately think the solution is for language servers to provide diagnostics and code actions together. Extending diagnostic-languageserver is an interesting idea, and I imagine other linters would benefit from a way to extract code actions from JSON output, but the output is very inconsistent between linters (even ESLint has several different formats). It would be a big task, and the community seems pretty split between [efm-langserver](https://github.com/mattn/efm-langserver) and diagnostic-languageserver.

Another option is chopping up diagnostic-languageserver to *only* provide ESLint diagnostics and code actions (and Prettier fixing). That would be more manageable, and while it wouldn't benefit other languages, it would also reduce the amount of boilerplate required to get up and running, which is one of the main reasons I made this plugin. 

VS Code and coc-eslint do indeed run non-conformant servers using the [Node.js API](https://eslint.org/docs/developer-guide/nodejs-api). Using `eslint_d` actually takes away a lot of the overhead and is just as fast in normal usage, and it wouldn't be impossible to use the same method to provide diagnostics, too, but at that point I'd rather write an actual language server. 

Going to close this for now because I just added a [Telescope snippet](https://github.com/jose-elias-alvarez/nvim-lsp-ts-utils/wiki/ESLint-code-actions-in-telescope.nvim) to the wiki, but I'm more than happy to continue the discussion in another issue. I'll start playing around with overriding `buf_request` and also see what I can do with diagnostic-languageserver.
